File: android/src/main/kotlin/com/example/pauza_screen_time/app_restriction/method_channel/RestrictionsMethodHandler.kt
package com.example.pauza_screen_time.app_restriction.method_channel

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.example.pauza_screen_time.core.PlatformConstants
import com.example.pauza_screen_time.app_restriction.schedule.RestrictionScheduledModeEntry
import com.example.pauza_screen_time.app_restriction.usecase.ConfigureShieldUseCase
import com.example.pauza_screen_time.app_restriction.usecase.LifecycleEventsUseCase
import com.example.pauza_screen_time.app_restriction.usecase.ManageModesUseCase
import com.example.pauza_screen_time.app_restriction.usecase.SessionEnforcementUseCase
import com.example.pauza_screen_time.core.MethodNames
import com.example.pauza_screen_time.core.PluginErrorHelper
import com.example.pauza_screen_time.permissions.PermissionHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class RestrictionsMethodHandler(
    private val contextProvider: () -> Context?,
    private val accessibilityStatusProvider: (Context) -> String = {
        PermissionHandler(it).checkPermission(PermissionHandler.ACCESSIBILITY_KEY)
    },
    private val lifecycleExecutor: ExecutorService = Executors.newSingleThreadExecutor(),
    private val resultPoster: ((() -> Unit) -> Unit)? = null,
) : MethodCallHandler {
    companion object {
        private const val ANDROID_ACCESSIBILITY_KEY = "android.accessibility"
        private const val FEATURE = "restrictions"
        private const val ACTIVE_SESSION_ERROR_MESSAGE =
            "A restriction session is already active. End the current session before starting a new one."
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                MethodNames.CONFIGURE_SHIELD -> handleConfigureShield(call, result)
                MethodNames.UPSERT_MODE -> handleUpsertMode(call, result)
                MethodNames.REMOVE_MODE -> handleRemoveMode(call, result)
                MethodNames.SET_MODES_ENABLED -> handleSetModesEnabled(call, result)
                MethodNames.GET_MODES_CONFIG -> handleGetModesConfig(result)
                MethodNames.IS_RESTRICTION_SESSION_ACTIVE_NOW -> handleIsRestrictionSessionActiveNow(result)
                MethodNames.PAUSE_ENFORCEMENT -> handlePauseEnforcement(call, result)
                MethodNames.RESUME_ENFORCEMENT -> handleResumeEnforcement(result)
                MethodNames.START_SESSION -> handleStartSession(call, result)
                MethodNames.END_SESSION -> handleEndSession(result)
                MethodNames.GET_PENDING_LIFECYCLE_EVENTS -> handleGetPendingLifecycleEvents(call, result)
                MethodNames.ACK_LIFECYCLE_EVENTS -> handleAckLifecycleEvents(call, result)
                MethodNames.GET_RESTRICTION_SESSION -> handleGetRestrictionSession(result)
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            PluginErrorHelper.internalFailure(
                result = result,
                feature = FEATURE,
                action = call.method,
                message = "Unexpected restriction error: ${e.message}",
                error = e,
            )
        }
    }

    fun dispose() {
        lifecycleExecutor.shutdown()
    }

    private fun postResult(action: () -> Unit) {
        val poster = resultPoster
        if (poster != null) {
            poster(action)
            return
        }
        Handler(Looper.getMainLooper()).post(action)
    }

    private fun handleConfigureShield(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.CONFIGURE_SHIELD)
        val configMap = call.arguments as? Map<String, Any?>
        if (configMap == null) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.CONFIGURE_SHIELD, "Shield configuration map is required")
            return
        }
        try {
            ConfigureShieldUseCase(context).execute(configMap)
            result.success(null)
        } catch (e: Exception) {
            internalFailure(result, MethodNames.CONFIGURE_SHIELD, "Failed to configure shield", e)
        }
    }

    private fun handleUpsertMode(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.UPSERT_MODE)
        if (emitRestrictionPreflightErrorIfAny(context, MethodNames.UPSERT_MODE, result)) return

        val payload = call.arguments as? Map<*, *>
        if (payload == null) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.UPSERT_MODE, "Missing or invalid mode payload")
            return
        }

        try {
            val mode = RestrictionScheduledModeEntry.fromMap(payload)
            ManageModesUseCase(context).upsertMode(mode.modeId, mode.blockedAppIds, mode.schedule)
            result.success(null)
        } catch (e: IllegalArgumentException) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.UPSERT_MODE, e.message ?: "Invalid mode payload")
        } catch (e: Exception) {
            internalFailure(result, MethodNames.UPSERT_MODE, "Failed to save mode", e)
        }
    }

    private fun handleRemoveMode(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.REMOVE_MODE)
        val modeId = ((call.arguments as? Map<*, *>)?.get("modeId") as? String)?.trim().orEmpty()
        if (modeId.isEmpty()) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.REMOVE_MODE, "Missing or invalid 'modeId' argument")
            return
        }
        try {
            ManageModesUseCase(context).removeMode(modeId)
            result.success(null)
        } catch (e: Exception) {
            internalFailure(result, MethodNames.REMOVE_MODE, "Failed to remove mode", e)
        }
    }

    private fun handleSetModesEnabled(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.SET_MODES_ENABLED)
        if (emitRestrictionPreflightErrorIfAny(context, MethodNames.SET_MODES_ENABLED, result)) return
        val enabled = (call.arguments as? Map<*, *>)?.get("enabled") as? Boolean
        if (enabled == null) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.SET_MODES_ENABLED, "Missing or invalid 'enabled' argument")
            return
        }
        try {
            ManageModesUseCase(context).setModesEnabled(enabled)
            result.success(null)
        } catch (e: Exception) {
            internalFailure(result, MethodNames.SET_MODES_ENABLED, "Failed to update modes toggle", e)
        }
    }

    private fun handleGetModesConfig(result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.GET_MODES_CONFIG)
        try {
            val config = ManageModesUseCase(context).getModesConfig()
            result.success(config.toChannelMap())
        } catch (e: Exception) {
            internalFailure(result, MethodNames.GET_MODES_CONFIG, "Failed to load modes config", e)
        }
    }

    private fun handleIsRestrictionSessionActiveNow(result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.IS_RESTRICTION_SESSION_ACTIVE_NOW)
        try {
            val isActive = SessionEnforcementUseCase(context).isRestrictionSessionActiveNow(areRestrictionPrerequisitesMet(context))
            result.success(isActive)
        } catch (e: Exception) {
            internalFailure(result, MethodNames.IS_RESTRICTION_SESSION_ACTIVE_NOW, "Failed to get restriction session active state", e)
        }
    }

    private fun handlePauseEnforcement(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.PAUSE_ENFORCEMENT)
        if (emitRestrictionPreflightErrorIfAny(context, MethodNames.PAUSE_ENFORCEMENT, result)) return

        val durationMs = call.argument<Number>("durationMs")?.toLong()
        if (durationMs == null || durationMs <= 0L) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.PAUSE_ENFORCEMENT, "Missing or invalid 'durationMs' argument")
            return
        }
        if (durationMs >= PlatformConstants.MAX_RELIABLE_PAUSE_DURATION_MS) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.PAUSE_ENFORCEMENT, "Pause duration must be less than 24 hours on Android")
            return
        }

        try {
            SessionEnforcementUseCase(context).pauseEnforcement(durationMs)
            result.success(null)
        } catch (e: IllegalStateException) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.PAUSE_ENFORCEMENT, e.message ?: "Invalid state for pause")
        } catch (e: Exception) {
            internalFailure(result, MethodNames.PAUSE_ENFORCEMENT, "Failed to pause restriction enforcement", e)
        }
    }

    private fun handleResumeEnforcement(result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.RESUME_ENFORCEMENT)
        if (emitRestrictionPreflightErrorIfAny(context, MethodNames.RESUME_ENFORCEMENT, result)) return
        try {
            SessionEnforcementUseCase(context).resumeEnforcement()
            result.success(null)
        } catch (e: Exception) {
            internalFailure(result, MethodNames.RESUME_ENFORCEMENT, "Failed to resume restriction enforcement", e)
        }
    }

    private fun handleStartSession(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.START_SESSION)
        if (emitRestrictionPreflightErrorIfAny(context, MethodNames.START_SESSION, result)) return
        val sessionUseCase = SessionEnforcementUseCase(context)
        if (emitActiveSessionErrorIfAny(sessionUseCase, MethodNames.START_SESSION, result)) return

        val payload = call.arguments as? Map<*, *>
        if (payload == null) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.START_SESSION, "Missing or invalid mode payload")
            return
        }

        try {
            val mode = RestrictionScheduledModeEntry.fromMap(payload)
            if (mode.blockedAppIds.isEmpty()) {
                PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.START_SESSION, "Mode requires non-empty 'blockedAppIds'")
                return
            }
            val durationMs = payload["durationMs"]?.let { rawDuration ->
                parseStartSessionDurationMs(rawDuration, result) ?: return
            }
            sessionUseCase.startSession(mode.modeId, mode.blockedAppIds, durationMs)
            result.success(null)
        } catch (e: IllegalArgumentException) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.START_SESSION, e.message ?: "Invalid mode payload")
        } catch (e: Exception) {
            internalFailure(result, MethodNames.START_SESSION, "Failed to start session", e)
        }
    }

    private fun handleEndSession(result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.END_SESSION)
        try {
            SessionEnforcementUseCase(context).endSession()
            result.success(null)
        } catch (e: Exception) {
            internalFailure(result, MethodNames.END_SESSION, "Failed to end session", e)
        }
    }

    private fun handleGetRestrictionSession(result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.GET_RESTRICTION_SESSION)
        try {
            val session = SessionEnforcementUseCase(context).getRestrictionSession()
            result.success(session.toChannelMap())
        } catch (e: Exception) {
            internalFailure(result, MethodNames.GET_RESTRICTION_SESSION, "Failed to get restriction session", e)
        }
    }

    private fun handleGetPendingLifecycleEvents(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.GET_PENDING_LIFECYCLE_EVENTS)
        val limit = ((call.arguments as? Map<*, *>)?.get("limit") as? Number)?.toInt() ?: PlatformConstants.DEFAULT_LIFECYCLE_EVENTS_LIMIT
        if (limit <= 0) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.GET_PENDING_LIFECYCLE_EVENTS, "Missing or invalid 'limit' argument")
            return
        }

        lifecycleExecutor.execute {
            try {
                val events = LifecycleEventsUseCase(context).getPendingLifecycleEvents(limit)
                postResult { result.success(events.map { it.toChannelMap() }) }
            } catch (e: Exception) {
                postResult { internalFailure(result, MethodNames.GET_PENDING_LIFECYCLE_EVENTS, "Failed to load pending lifecycle events", e) }
            }
        }
    }

    private fun handleAckLifecycleEvents(call: MethodCall, result: Result) {
        val context = contextProvider() ?: return noContext(result, MethodNames.ACK_LIFECYCLE_EVENTS)
        val throughEventId = ((call.arguments as? Map<*, *>)?.get("throughEventId") as? String)?.trim().orEmpty()
        if (throughEventId.isEmpty()) {
            PluginErrorHelper.invalidArgument(result, FEATURE, MethodNames.ACK_LIFECYCLE_EVENTS, "Missing or invalid 'throughEventId' argument")
            return
        }

        lifecycleExecutor.execute {
            try {
                LifecycleEventsUseCase(context).ackLifecycleEventsThrough(throughEventId)
                postResult { result.success(null) }
            } catch (e: Exception) {
                postResult { internalFailure(result, MethodNames.ACK_LIFECYCLE_EVENTS, "Failed to acknowledge lifecycle events", e) }
            }
        }
    }

    private fun areRestrictionPrerequisitesMet(context: Context): Boolean {
        return getMissingPrerequisites(context).isEmpty()
    }

    private fun getMissingPrerequisites(context: Context): List<String> {
        val accessibilityStatus = accessibilityStatusProvider(context)
        if (accessibilityStatus == PermissionHandler.STATUS_GRANTED) {
            return emptyList()
        }
        return listOf(ANDROID_ACCESSIBILITY_KEY)
    }

    private fun emitRestrictionPreflightErrorIfAny(context: Context, action: String, result: Result): Boolean {
        val accessibilityStatus = accessibilityStatusProvider(context)
        if (accessibilityStatus == PermissionHandler.STATUS_GRANTED) {
            return false
        }

        PluginErrorHelper.missingPermission(
            result = result,
            feature = FEATURE,
            action = action,
            message = "Accessibility permission is required for restrictions",
            missing = listOf(ANDROID_ACCESSIBILITY_KEY),
            status = mapOf("androidAccessibilityStatus" to accessibilityStatus),
        )
        return true
    }

    private fun emitActiveSessionErrorIfAny(
        sessionUseCase: SessionEnforcementUseCase,
        action: String,
        result: Result,
    ): Boolean {
        if (!sessionUseCase.hasActiveSession()) {
            return false
        }
        PluginErrorHelper.invalidArgument(result, FEATURE, action, ACTIVE_SESSION_ERROR_MESSAGE)
        return true
    }

    private fun parseStartSessionDurationMs(rawDurationMs: Any?, result: Result): Long? {
        val durationMs = (rawDurationMs as? Number)?.toLong()
        if (durationMs == null || durationMs <= 0L) {
            PluginErrorHelper.invalidArgument(
                result,
                FEATURE,
                MethodNames.START_SESSION,
                "Missing or invalid 'durationMs' argument",
            )
            return null
        }
        if (durationMs >= PlatformConstants.MAX_RELIABLE_PAUSE_DURATION_MS) {
            PluginErrorHelper.invalidArgument(
                result,
                FEATURE,
                MethodNames.START_SESSION,
                "Session duration must be less than 24 hours on Android",
            )
            return null
        }
        return durationMs
    }

    private fun noContext(result: Result, action: String) {
        PluginErrorHelper.internalFailure(result, FEATURE, action, "Application context is not available")
    }

    private fun internalFailure(result: Result, action: String, message: String, e: Exception) {
        PluginErrorHelper.internalFailure(result, FEATURE, action, "$message: ${e.message}", e.message)
    }
}

File: ios/Classes/AppRestriction/MethodChannel/RestrictionsMethodHandler.swift
import Flutter
import FamilyControls
import Foundation

final class RestrictionsMethodHandler {
    private static let iosFamilyControlsKey = "ios.familyControls"
    private static let featureRestrictions = "restrictions"
    private static let platformIOS = "ios"
    private static let maxReliablePauseDurationMs: Int64 = 24 * 60 * 60 * 1000
    private let lifecycleQueue = DispatchQueue(label: "pauza.restrictions.lifecycle.queue")

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case MethodNames.configureShield:
            handleConfigureShield(call: call, result: result)
        case MethodNames.upsertMode:
            handleUpsertMode(call: call, result: result)
        case MethodNames.removeMode:
            handleRemoveMode(call: call, result: result)
        case MethodNames.setModesEnabled:
            handleSetModesEnabled(call: call, result: result)
        case MethodNames.getModesConfig:
            handleGetModesConfig(result: result)
        case MethodNames.isRestrictionSessionActiveNow:
            handleIsRestrictionSessionActiveNow(result: result)
        case MethodNames.pauseEnforcement:
            handlePauseEnforcement(call: call, result: result)
        case MethodNames.resumeEnforcement:
            handleResumeEnforcement(result: result)
        case MethodNames.startSession:
            handleStartSession(call: call, result: result)
        case MethodNames.endSession:
            handleEndSession(result: result)
        case MethodNames.getPendingLifecycleEvents:
            handleGetPendingLifecycleEvents(call: call, result: result)
        case MethodNames.ackLifecycleEvents:
            handleAckLifecycleEvents(call: call, result: result)
        case MethodNames.getRestrictionSession:
            handleGetRestrictionSession(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleConfigureShield(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let configuration = call.arguments as? [String: Any] else {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.configureShield,
                message: PluginErrorMessage.missingShieldConfiguration
            ))
            return
        }
        if let error = ConfigureShieldUseCase.execute(configuration: configuration) {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleIsRestrictionSessionActiveNow(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(false)
            return
        }
        let isPrerequisitesMet = restrictionMissingPrerequisites().isEmpty
        let isActive = SessionEnforcementUseCase.isRestrictionSessionActiveNow(isPrerequisitesMet: isPrerequisitesMet)
        result(isActive)
    }

    private func handlePauseEnforcement(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(
                feature: Self.featureRestrictions,
                action: MethodNames.pauseEnforcement,
                message: PluginErrorMessage.restrictionsUnsupported
            ))
            return
        }
        if let preflightError = restrictionPreflightError(action: MethodNames.pauseEnforcement) {
            result(preflightError)
            return
        }
        guard let args = call.arguments as? [String: Any],
              let durationValue = args["durationMs"] as? NSNumber else {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.pauseEnforcement,
                message: "Missing or invalid 'durationMs' argument"
            ))
            return
        }
        let durationMs = durationValue.int64Value
        if durationMs <= 0 {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.pauseEnforcement,
                message: "Missing or invalid 'durationMs' argument"
            ))
            return
        }
        if durationMs >= Self.maxReliablePauseDurationMs {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.pauseEnforcement,
                message: PluginErrorMessage.pauseTooLong
            ))
            return
        }
        if RestrictionStateStore.loadPausedUntilEpochMs() > 0 {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.pauseEnforcement,
                message: "Restriction enforcement is already paused"
            ))
            return
        }
        if let error = SessionEnforcementUseCase.pauseEnforcement(durationMs: durationMs) {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleResumeEnforcement(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(
                feature: Self.featureRestrictions,
                action: MethodNames.resumeEnforcement,
                message: PluginErrorMessage.restrictionsUnsupported
            ))
            return
        }
        if let preflightError = restrictionPreflightError(action: MethodNames.resumeEnforcement) {
            result(preflightError)
            return
        }
        if let error = SessionEnforcementUseCase.resumeEnforcement() {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleStartSession(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(
                feature: Self.featureRestrictions,
                action: MethodNames.startSession,
                message: PluginErrorMessage.restrictionsUnsupported
            ))
            return
        }
        if let preflightError = restrictionPreflightError(action: MethodNames.startSession) {
            result(preflightError)
            return
        }
        guard let args = call.arguments as? [String: Any] else {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.startSession,
                message: "Missing or invalid mode payload"
            ))
            return
        }

        let modeId = (args["modeId"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let blockedAppIdsRaw = args["blockedAppIds"] as? [Any] ?? []
        let blockedAppIds = blockedAppIdsRaw.compactMap { value -> String? in
            guard let raw = value as? String else { return nil }
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            return trimmed.isEmpty ? nil : trimmed
        }
        if modeId.isEmpty || blockedAppIds.isEmpty {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.startSession,
                message: "Mode requires non-empty 'modeId' and 'blockedAppIds'"
            ))
            return
        }
        if let error = SessionEnforcementUseCase.startSession(modeId: modeId, blockedAppIds: blockedAppIds) {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleEndSession(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(
                feature: Self.featureRestrictions,
                action: MethodNames.endSession,
                message: PluginErrorMessage.restrictionsUnsupported
            ))
            return
        }
        if let error = SessionEnforcementUseCase.endSession() {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleGetPendingLifecycleEvents(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result([])
            return
        }
        let args = call.arguments as? [String: Any]
        let limit = (args?["limit"] as? NSNumber)?.intValue ?? 200
        if limit <= 0 {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.getPendingLifecycleEvents,
                message: "Missing or invalid 'limit' argument"
            ))
            return
        }
        lifecycleQueue.async {
            let events = RestrictionStateStore.loadPendingLifecycleEvents(limit: limit)
            DispatchQueue.main.async {
                result(events.map { $0.toChannelMap() })
            }
        }
    }

    private func handleAckLifecycleEvents(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(nil)
            return
        }
        guard let args = call.arguments as? [String: Any],
              let throughEventIdRaw = args["throughEventId"] as? String else {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.ackLifecycleEvents,
                message: "Missing or invalid 'throughEventId' argument"
            ))
            return
        }
        let throughEventId = throughEventIdRaw.trimmingCharacters(in: .whitespacesAndNewlines)
        if throughEventId.isEmpty {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.ackLifecycleEvents,
                message: "Missing or invalid 'throughEventId' argument"
            ))
            return
        }
        lifecycleQueue.async {
            let error = LifecycleEventsUseCase.ackLifecycleEvents(throughEventId: throughEventId)
            DispatchQueue.main.async {
                if let error {
                    result(error)
                } else {
                    result(nil)
                }
            }
        }
    }

    private func handleGetRestrictionSession(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(RestrictionSessionSnapshot(
                isScheduleEnabled: false,
                isInScheduleNow: false,
                pausedUntilEpochMs: nil,
                activeMode: nil,
                activeModeSource: .none,
                currentSessionEvents: []
            ).toChannelMap())
            return
        }
        let session = SessionEnforcementUseCase.getRestrictionSession()
        result(session.toChannelMap())
    }

    private func handleUpsertMode(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(
                feature: Self.featureRestrictions,
                action: MethodNames.upsertMode,
                message: PluginErrorMessage.restrictionsUnsupported
            ))
            return
        }
        if let preflightError = restrictionPreflightError(action: MethodNames.upsertMode) {
            result(preflightError)
            return
        }
        guard let args = call.arguments as? [String: Any],
              let mode = RestrictionScheduledMode(channelMap: args) else {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.upsertMode,
                message: "Missing or invalid mode payload"
            ))
            return
        }
        if let error = ManageModesUseCase.upsertMode(mode: mode) {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleRemoveMode(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(
                feature: Self.featureRestrictions,
                action: MethodNames.removeMode,
                message: PluginErrorMessage.restrictionsUnsupported
            ))
            return
        }
        guard let args = call.arguments as? [String: Any],
              let modeIdRaw = args["modeId"] as? String else {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.removeMode,
                message: "Missing or invalid 'modeId' argument"
            ))
            return
        }
        let modeId = modeIdRaw.trimmingCharacters(in: .whitespacesAndNewlines)
        if modeId.isEmpty {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.removeMode,
                message: "Missing or invalid 'modeId' argument"
            ))
            return
        }
        if let error = ManageModesUseCase.removeMode(modeId: modeId) {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleSetModesEnabled(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(PluginErrors.unsupported(
                feature: Self.featureRestrictions,
                action: MethodNames.setModesEnabled,
                message: PluginErrorMessage.restrictionsUnsupported
            ))
            return
        }
        if let preflightError = restrictionPreflightError(action: MethodNames.setModesEnabled) {
            result(preflightError)
            return
        }
        guard let args = call.arguments as? [String: Any],
              let enabled = args["enabled"] as? Bool else {
            result(PluginErrors.invalidArguments(
                feature: Self.featureRestrictions,
                action: MethodNames.setModesEnabled,
                message: "Missing or invalid 'enabled' argument"
            ))
            return
        }
        if let error = ManageModesUseCase.setModesEnabled(enabled: enabled) {
            result(error)
        } else {
            result(nil)
        }
    }

    private func handleGetModesConfig(result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            result(RestrictionScheduledModesConfig(enabled: false, modes: []).toChannelMap())
            return
        }
        let config = ManageModesUseCase.getModesConfig()
        result(config.toChannelMap())
    }

    @available(iOS 16.0, *)
    private func restrictionMissingPrerequisites() -> [String] {
        let isAuthorized = AuthorizationCenter.shared.authorizationStatus == .approved
        return isAuthorized ? [] : [Self.iosFamilyControlsKey]
    }

    @available(iOS 16.0, *)
    private func restrictionPreflightError(action: String) -> FlutterError? {
        let authorizationStatus = AuthorizationCenter.shared.authorizationStatus
        if authorizationStatus == .approved {
            return nil
        }
        let details: [String: Any] = [
            "feature": Self.featureRestrictions,
            "action": action,
            "platform": Self.platformIOS,
            "missing": [Self.iosFamilyControlsKey],
            "status": [
                "iosAuthorizationStatus": iosAuthorizationStatusKey(authorizationStatus)
            ],
        ]
        switch authorizationStatus {
        case .notDetermined:
            return PluginErrors.missingPermission(
                feature: Self.featureRestrictions,
                action: action,
                message: "Screen Time authorization is required for restrictions",
                missing: details["missing"] as? [String],
                status: details["status"] as? [String: Any]
            )
        case .denied:
            return PluginErrors.permissionDenied(
                feature: Self.featureRestrictions,
                action: action,
                message: "Screen Time authorization was denied",
                missing: details["missing"] as? [String],
                status: details["status"] as? [String: Any]
            )
        @unknown default:
            return PluginErrors.systemRestricted(
                feature: Self.featureRestrictions,
                action: action,
                message: "Screen Time authorization is unavailable",
                missing: details["missing"] as? [String],
                status: details["status"] as? [String: Any]
            )
        }
    }

    @available(iOS 16.0, *)
    private func iosAuthorizationStatusKey(_ status: AuthorizationStatus) -> String {
        switch status {
        case .approved: return "approved"
        case .denied: return "denied"
        case .notDetermined: return "notDetermined"
        @unknown default: return "unknown"
        }
    }
}
